### ISSUE TITLE
False positive for `use_decorated_box`

### ISSUE THREAD

**Issue by ookami-kb** [dart-lang/linter#3239]
**Describe the issue**

`Container` has nullable `decoration` parameter, so construction like this is totally valid:

```dart
Container(
  decoration: _counter.isEven
      ? const BoxDecoration(color: Colors.green)
      : null,
  child: const Text(
    'You have pushed the button this many times:',
  ),
),
```

But with `use_decorated_box` lint rule enable, it suggests using `DecoratedBox` instead. But it's not possible since `DecoratedBox` doesn't accept nullable `Decoration?`.

**To Reproduce**

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

const Decoration? decoration = null;

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Container(
        decoration: decoration,
        child: const Center(
          child: Text('Hello World'),
        ),
      ),
    );
  }
}
```

**Expected behavior**

`use_decorated_box` shouldn't trigger if nullable `Decoration?` is passed as a parameter to `Container`.

**Comment by minhqdao** [comment]
You're right, a `DecoratedBox` doesn't accept a nullable `Decoration?`, which is why you're using a `Container` as a fallback, that does. In my opinion, the linting rule is actually correct, but a `DecoratedBox` should accept a `Decoration?` instead of a `Decoration` to be able to deal with cases such as yours.

As I also implemented the original rule, I can also try to fix it, if that is generally desired. But again, I find it a bit strange (actually, I find it wrong) to loosen the rule on `Container` just because a `DecoratedBox` cannot handle `null`.

**Comment by ookami-kb** [comment]
Well, as for me, `DecoratedBox` without `decoration` looks really weird and could lead to even more non-obvious behavior. E.g. construction like this would definitely be weird:

```dart
DecoratedBox(child: Text('Hello World'))
```

From that perspective, I would say that fixing `use_decorated_box` makes more sense: it shouldn't advise switching to `DecoratedBox` if you **cannot** switch to it, whatever the actual reason is.

**Comment by minhqdao** [comment]
I don't think that an optional `decoration` would be weird at all. It just resembles a `Container`, where all parameters are optional to enable optional injection (the primary motivation for this fix).

But I entirely agree with this: 👇🏽

> it shouldn't advise switching to `DecoratedBox` if you **cannot** switch to it, whatever the actual reason is.

**Comment by minhqdao** [comment]
We'd need to "fix" the newly introduced `use_colored_box`, too. 🙈

**Comment by goderbauer** [comment]
Making the decoration of DecoratedBox nullable seems strange. If you don't want the decoration, don't include the DecoratedBox in the first place.

One could argue that

```dart
return Container(
  decoration: _counter.isEven
      ? const BoxDecoration(color: Colors.green)
      : null,
  child: const Text(
    'You have pushed the button this many times:',
  ),
),
```

should actually be changed to something like:

```dart
Widget child = const Text(
    'You have pushed the button this many times:',
  );
if (_counter.isEven) {
  child = DecoratedBox(
    decoration: const BoxDecoration(color: Colors.green),
    child: child,
  );
}
return child;
```

That's essentially what `Container` does internally as well.

It's certainly not as ergonomic as just using a `Container`, but it appears to be in the spirit of the `use_decorated_box` lint, which is essentially about avoiding unnecessary widgets - and if _counter is not even, you don't need the DecoratedBox.

Long story short, I don't think this is a false positive. But it could be discussed if for ergonomic reasons the lint shouldn't fire in this particular case (although I think that's somewhat against the current spirit of the lint).

**Comment by ookami-kb** [comment]
> That's essentially what Container does internally as well.

But that's the point. This makes the hint suggested by `use_decorated_box`:
- not possible to implement in a "straightforward" way – by replacing `Container` with `DecoratedBox`;
- useless to implement by the way you suggest – since `Container` would do the same.

In my opinion, that makes it a false positive.

**Comment by goderbauer** [comment]
> useless to implement in the way you suggest – since Container would do the same.

No, container does much more than that - all of that logic gets cut out when you DecoratedBox directly.

**Comment by ookami-kb** [comment]
If you pass only the decoration, the result would be pretty much the same. Ok, there will be several additional `if`s, but it definitely won't affect performance to worry about `Container` vs conditional-decorated-box-or-just-a-child.

**Comment by goderbauer** [comment]
> but it definitely won't affect performance to worry about Container vs conditional-decorated-box-or-just-a-child.

Then don't use the lint if you don't care about that difference. The premise of the lint is that people care about that difference.

**Comment by ookami-kb** [comment]
I care about that difference when it makes sense.

`const DecoratedBox(decoration: decoration, child: child)` vs `Container(decoration: decoration, child: child)` makes sense, because `DecoratedBox` has a `const` constructor.

`Container(decoration: decoration, child: child)` vs `decoration != null ? DecoratedBox(decoration: decoration!, child: child) : child` doesn't make that much sense to me since the result would be *almost* the same, but the readability is much worse.

**Comment by thaihuynhxyz** [comment]
I think the lint still makes sense when `Container` has only a `Decoration`, we can use the condition to use a `DecoratedBox` or only `child` is enough.

```dart
class MyBox extends StatelessWidget {

  const MyBox({this.decoration, required this.child, Key? key})
      : super(key: key);
  
  final Widget child;
  final Decoration? decoration;

  @override
  Widget build(BuildContext context) {
    return decoration != null
        ? DecoratedBox(decoration: decoration!, child: child)
        : child;
  }
}
```

**Comment by srawlins** [comment]
I'll close this as not-a-bug, as per @goderbauer 's reasons above.
